### ISSUE TITLE
Keras-Tensorflow does not produce reproducible results on GPU

### ISSUE THREAD

**Issue by emirceyani** [keras-team/keras#7937]
Hello folks,

My aim to ask this question is that, there is no unified answer for getting producible results with Keras-Tensorflow on GPU. So, If you'd like to answer, that'd be better to all of us.

I tried the following:

1) Import numpy. Seed it. Then import all other libraries

2) https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development

What is the correct way of doing this?

Thanks in advance

**Comment by primeMover2011** [comment]
What is the problem you encounter? Did you seed random and tensorflow as well?
Could you share a little code that shows how to reproduce your issue?

**Comment by emirceyani** [comment]
I cannot share the full code but I shared my training script as follows:

https://gist.github.com/emirceyani/3a07b80d53158564e1f7c90ea4922785

preprocessing is in pure Python and there are custom blocks too

**Comment by primeMover2011** [comment]
Did you try to set the seed on the DropOut Layer to 0?

**Comment by emirceyani** [comment]
Hello, I tried the 2nd solution again and it worked on CPU. I haven't tried
it on GPU yet.


2017-09-21 17:04 GMT+03:00 HerbertD <notifications@github.com>:

**Comment by grafael** [comment]
If you are using **GPU** and **cudnn** there are chances that this will not work. Some of **cudnn** algorithms implementations are *non deterministic*.

**Comment by rpryzant** [comment]
If it means anything, I've been able to get reproducible results on a gpu with keras by seeding python, numpy _and_ tensorflow:

```
tf.set_random_seed(seed)
np.random.seed(seed)
random.seed(seed)
```

**Comment by fuzunyun** [comment]
Hello, I meet the save question,and I also tried the method recommended by https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development
unfortunately,it dont work. how to reproduce the same result?

**Comment by Vijetha1** [comment]
I too am looking for an answer to this question. Is there currently no way to reproduce the results when trained on GPU using Keras with tensorflow back-end?

**Comment by JuntingGuo** [comment]
I tried on several ways and for now the only time I succeeded in having reproducible results is when I forced the script to run in CPU and explicitly selected the same CPU device using `with tf.device('/cpu:0'):` . It seems that as long as you use GPU for training you will not have the same results.  I saw in other issue threads that someone mentioned he made it by updating tensorflow to 1.5.0. But I couldn't try it.

**Comment by ymodak** [comment]
Please use the latest version of TensorFlow and test again. Feel free to reopen the issue if it still persists.
Thanks!

**Comment by mhaghighat** [comment]
The results are still not reproducible with tensorflow-gpu==2.2. 
They are reproducible on CPU though.

**Comment by mhaghighat** [comment]
I could get reproducible results using the [TensorFlow Determinism](https://github.com/NVIDIA/framework-determinism) package.
